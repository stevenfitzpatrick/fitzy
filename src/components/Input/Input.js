import classnames from 'classnames';
import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';

import { isBoolean } from '../../utils/formUtils';
import {
  FitzyInput,
  FitzyLabel,
  FormError,
  FormField,
  InputBar
} from './Input.styled';

class Input extends PureComponent {
  getEventValue = (type, e) =>
    isBoolean(type) ? e.target.checked : e.target.value;

  handleChange = e =>
    this.props.onChange &&
    this.props.onChange(e, this.getEventValue(this.props.type, e));

  handleBlur = e => {
    this.props.onBlur &&
      this.props.onBlur(e, this.getEventValue(this.props.type, e));
  };

  renderHint() {
    const { hint } = this.props;

    if (!hint) {
      return null;
    }

    return <p>{hint}</p>;
  }

  renderWarning() {
    const { warning } = this.props;

    if (!warning) {
      return null;
    }

    return <FormError>{warning}</FormError>;
  }

  renderLabel(inputId, labelId) {
    const { label } = this.props;

    if (!label) {
      return null;
    }

    return (
      <FitzyLabel id={labelId} htmlFor={inputId}>
        {label}
      </FitzyLabel>
    );
  }

  renderInput(inputId, labelId) {
    const {
      type,
      name,
      value,
      placeholder,
      required,
      disabled,
      autoFocus,
      spellCheck,
      autoComplete,
      innerRef
    } = this.props;

    const classNames = classnames('');

    return (
      <FitzyInput
        name={name}
        className={classNames}
        value={value}
        placeholder={placeholder}
        type={type}
        id={inputId}
        aria-labelledby={labelId}
        required={required}
        disabled={disabled}
        autoFocus={autoFocus}
        spellCheck={spellCheck}
        autoComplete={autoComplete}
        innerRef={innerRef}
        onChange={this.handleChange}
        onBlur={this.handleBlur}
      />
    );
  }

  render() {
    const { id, name, className, warning } = this.props;
    const inputId = id || name;
    const labelId = `label-${inputId}`;

    return (
      <FormField role="presentation" className={className} warning={warning}>
        {this.renderInput(inputId, labelId)}
        <InputBar />
        {this.renderLabel(inputId, labelId)}
        {this.renderHint()}
        {this.renderWarning()}
      </FormField>
    );
  }
}

Input.propTypes = {
  /**
   * onChange Handler sending updated value
   */
  onChange: PropTypes.func.isRequired,
  /**
   * onBlur Handler sending updated value
   */
  onBlur: PropTypes.func,
  /**
   * Current Value of the input
   */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.bool
  ]),
  /**
   * Set input type
   */
  type: PropTypes.oneOf([
    'checkbox',
    'text',
    'url',
    'email',
    'color',
    'date',
    'number',
    'password',
    'range',
    'search',
    'tel',
    'textarea'
  ]),
  /**
   * Set Label for input field
   */
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  /**
   * Set name for input field
   */
  name: PropTypes.string,
  /**
   * Set placeholder for input field
   */
  placeholder: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * Set hint for input field
   */
  hint: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * Set id for input field
   */
  id: PropTypes.string,
  /**
   * Mark the field as required
   */
  required: PropTypes.bool,
  /**
   * Mark the field as disabled
   */
  disabled: PropTypes.bool,
  /**
   * Add SpellCheck functionality to input
   */
  spellCheck: PropTypes.bool,
  /**
   * Add Auto Focus functionality to input
   */
  autoFocus: PropTypes.bool,
  /**
   * Add Auto Complete aka autofill functionality to input
   */
  autoComplete: PropTypes.string,
  /**
   * Add Error Message below search field
   */
  warning: PropTypes.node,
  /**
   * Classname for css class for input container
   */
  className: PropTypes.string,
  /**
   * Pass down a ref for the Input
   */
  innerRef: PropTypes.oneOfType([PropTypes.object, PropTypes.func])
};

Input.defaultProps = {
  type: 'text',
  required: false,
  placeholder: null,
  hint: '',
  value: '',
  disabled: false,
  spellCheck: null,
  autoFocus: null,
  autoComplete: null,
  warning: null,
  className: ''
};

export default Input;
