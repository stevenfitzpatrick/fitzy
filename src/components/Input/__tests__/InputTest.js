import React from 'react';
import {
  mountWithTheme,
  renderWithTheme,
  shallowWithTheme
} from '@sfitzpatrick/enzyme-context-helpers';

import Input from '../Input';

describe('Input', () => {
  let wrapper, shallowWrapper;
  let input;

  const onChangeSpy = jest.fn().mockName('onChange');
  const onBlurSpy = jest.fn().mockName('onBlur');

  const defaultProps = {
    type: 'text',
    value: 'test',
    name: 'name',
    placeholder: 'Enter Name',
    onChange: onChangeSpy,
    onBlur: onBlurSpy
  };

  const funkyProps = {
    ...defaultProps,
    spellCheck: false,
    autoComplete: 'username',
    autoFocus: true
  };

  beforeEach(() => {
    wrapper = mountWithTheme(<Input type="text" {...defaultProps} />);
    shallowWrapper = shallowWithTheme(<Input type="text" {...defaultProps} />);
    input = wrapper.find('input');
  });

  afterEach(() => {
    onChangeSpy.mockReset();
  });

  test('should match snapshot with label', () => {
    const tree = renderWithTheme(
      <Input type="text" {...defaultProps} label="My Name" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should render Input field with type text', () => {
    expect(wrapper.find('Input')).toHaveLength(1);
    expect(shallowWrapper.dive().is('label')).toBeTruthy();
    expect(input.props().type).toEqual('text');
  });

  test('should handle default props', () => {
    expect(input.props().placeholder).toEqual('Enter Name');
    expect(input.props().disabled).toEqual(false);
    expect(input.props().type).toEqual('text');
  });

  test('should set id based on name if no id entered', () => {
    expect(input.props().id).toEqual(defaultProps.name);
  });

  test('should call onChange when text entered', () => {
    expect(onChangeSpy.mock.calls).toHaveLength(0);
    input.simulate('change', { target: { value: 'new' } });
    expect(onChangeSpy.mock.calls).toHaveLength(1);
    expect(onChangeSpy.mock.calls[0][1]).toBe('new');
  });

  test('should call onBlur when leaving input', () => {
    expect(onBlurSpy.mock.calls).toHaveLength(0);
    input.simulate('blur', { target: { value: 'new' } });
    expect(onBlurSpy.mock.calls).toHaveLength(1);
  });

  describe('with Label', () => {
    beforeEach(() => wrapper.setProps({ label: 'Name:' }));

    test('should render label when set', () => {
      expect(wrapper.find('label')).toHaveLength(2);
    });

    test('should set text and for correctly', () => {
      const inputLabel = wrapper.find('label').last();
      expect(inputLabel.text()).toEqual('Name:');
      expect(inputLabel.props().htmlFor).toEqual('name');
    });
  });

  describe('with Hint', () => {
    let hint;

    beforeEach(() => {
      wrapper.setProps({ hint: 'Only Unique' });
      hint = wrapper.find('p');
    });

    test('should render hint when set', () => {
      expect(hint).toHaveLength(1);
      expect(hint.containsMatchingElement(<p>Only Unique</p>)).toEqual(true);
    });

    test('should set text correctly', () => {
      expect(hint.text()).toEqual('Only Unique');
    });
  });

  describe('with Funky Props', () => {
    test('should display all extra funky props', () => {
      wrapper.setProps({ ...funkyProps });
      const funkyInput = wrapper.find('input');
      expect(Object.keys(funkyInput.props())).toEqual(
        expect.arrayContaining(Object.keys(funkyProps))
      );
    });
  });
});
