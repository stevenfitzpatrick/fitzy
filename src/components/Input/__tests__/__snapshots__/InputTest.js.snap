// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should match snapshot with label 1`] = `
.c0 {
  position: relative;
  display: block;
  margin-bottom: 2.25rem;
  padding-top: 0.75rem;
}

.c2 {
  position: relative;
  display: block;
  width: 100%;
}

.c2::before,
.c2::after {
  content: '';
  height: 3px;
  width: 0;
  bottom: 0;
  position: absolute;
  background-color: hsl(152,62%,63%);
  -webkit-transition: 0.2s ease all;
  transition: 0.2s ease all;
}

.c2::before {
  left: 50%;
}

.c2::after {
  right: 50%;
}

.c3 {
  position: absolute;
  display: inline-block;
  max-width: 100%;
  font-size: 0.875rem;
  line-height: 1;
  color: hsl(152,62%,53%);
  top: 1rem;
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
  pointer-events: none;
  opacity: 0;
  -webkit-transform: translateY(0);
  -ms-transform: translateY(0);
  transform: translateY(0);
}

.c1 {
  position: relative;
  display: block;
  padding: 0.5rem 0;
  width: 100%;
  line-height: 1.25rem;
  font-size: 1rem;
  font-weight: normal;
  font-family: inherit;
  background-color: transparent;
  border: none;
  border-bottom: 1px solid hsl(152,62%,53%);
  color: #fff;
  -webkit-transition: border 0.3s ease-in-out;
  transition: border 0.3s ease-in-out;
}

.c1 ~ label {
  -webkit-transform: translateY(-1rem);
  -ms-transform: translateY(-1rem);
  transform: translateY(-1rem);
  opacity: 1;
}

.c1::-webkit-input-placeholder {
  color: #b0c1d4;
}

.c1::-moz-placeholder {
  color: #b0c1d4;
}

.c1:-ms-input-placeholder {
  color: #b0c1d4;
}

.c1::placeholder {
  color: #b0c1d4;
}

.c1:-webkit-autofill {
  box-shadow: 0 0 0 30px #172025 inset;
  -webkit-text-fill-color: #ffffff !important;
  caret-color: #b0c1d4;
}

.c1:-webkit-autofill::-webkit-input-placeholder {
  color: #b0c1d4;
}

.c1:-webkit-autofill::-moz-placeholder {
  color: #b0c1d4;
}

.c1:-webkit-autofill:-ms-input-placeholder {
  color: #b0c1d4;
}

.c1:-webkit-autofill::placeholder {
  color: #b0c1d4;
}

.c1:-webkit-autofill ~ label {
  -webkit-transform: translateY(-1rem);
  -ms-transform: translateY(-1rem);
  transform: translateY(-1rem);
  opacity: 1;
}

.c1:focus {
  border-bottom: 1px solid hsl(152,62%,63%);
  outline: none;
}

.c1:focus ~ label {
  color: hsl(152,62%,63%);
  -webkit-transform: translateY(-1rem);
  -ms-transform: translateY(-1rem);
  transform: translateY(-1rem);
  opacity: 1;
}

.c1:focus ~ span::before,
.c1:focus ~ span::after {
  width: 50%;
}

<label
  className="c0"
  role="presentation"
>
  <input
    aria-labelledby="label-name"
    autoComplete={null}
    autoFocus={null}
    className="c1"
    disabled={false}
    id="name"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Enter Name"
    required={false}
    spellCheck={null}
    type="text"
    value="test"
  />
  <span
    className="c2"
  />
  <label
    className="c3"
    htmlFor="name"
    id="label-name"
  >
    My Name
  </label>
</label>
`;
